src/hotspot/os/windows/attachListener_windows.cpp

```diff
@@ -55,8 +62,154 @@
 // this wasn't worth worrying about.
 
 
-// forward reference
-class Win32AttachOperation;
+class PipeChannel : public AttachOperation::RequestReader, public AttachOperation::ReplyWriter {
+private:
+  HANDLE _hPipe;
+public:
+  PipeChannel() : _hPipe(INVALID_HANDLE_VALUE) {}
+  ~PipeChannel() {
+    close();
+  }
+
+  bool opened() const {
+    return _hPipe != INVALID_HANDLE_VALUE;
+  }
+
+  bool open(const char* pipe, bool write_only) {
+    _hPipe = ::CreateFile(pipe,
+                          GENERIC_WRITE | (write_only ? 0 : GENERIC_READ),
+                          0,              // no sharing
+                          nullptr,        // default security attributes
+                          OPEN_EXISTING,  // opens existing pipe
+                          0,              // default attributes
+                          nullptr);       // no template file
+    if (_hPipe == INVALID_HANDLE_VALUE) {
+      log_error(attach)("could not open (%d) pipe %s", GetLastError(), pipe);
+      return false;
+    }
+    return true;
+  }
+
+  void close() {
+    if (opened()) {
+      CloseHandle(_hPipe);
+      _hPipe = INVALID_HANDLE_VALUE;
+    }
+  }
+
+  // RequestReader
+  int read(void* buffer, int size) override {
+    assert(opened(), "must be");
+    DWORD nread;
+    BOOL fSuccess= ReadFile(_hPipe,
+                            buffer,
+                            size,
+                            &nread,
+                            nullptr);   // not overlapped
```

Comment:
Nit: Missed space after `fsuccess` at line 104.
       Also, there is not cast `(DWORD)size` at line 106 as it is done in the `WriteFile()` call at line 118.

Reply:
Fixed both indent and cast